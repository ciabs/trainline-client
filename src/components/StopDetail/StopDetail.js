import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import FontAwesome  from 'react-fontawesome';

import {Circle, CircleMiddle, Container, LeftCell, Location, RightCell, TrainStateStyled} from './StopDetail.styles';
import config from '../../config';
import {getTrainInTheMiddle, getTrainPositon, getTrainState} from '../../helpers/trainDetailsHelper';

const StopDetail = ({stop, nextStop}) => {
  const time = stop.arrival && stop.arrival.scheduled ?
    moment(stop.arrival.scheduled.scheduledTime).format('HH:mm') :
    moment(stop.departure.scheduled.scheduledTime).format('HH:mm');

  const trainPosition = getTrainPositon(stop);
  const trainState = getTrainState(stop);
  const isLastStop = stop.departure && stop.departure.notApplicable;
  const inTheMidle = getTrainInTheMiddle(stop, nextStop);
  const isLocationPassed =
    stop.departure &&
    stop.departure.realTime &&
    stop.departure.realTime.realTimeServiceInfo &&
    stop.departure.realTime.realTimeServiceInfo.hasDeparted;


  const renderCircle = (
    <Circle trainPosition={trainPosition}>
      {
        trainPosition === config.trainPosition.TRAIN &&
        <FontAwesome
          name='train'
        />
      }
    </Circle>
  );

  const renderMiddleCircle = inTheMidle && (
    <CircleMiddle trainPosition={config.trainPosition.TRAIN}>
      <FontAwesome
        name='train'
      />
    </CircleMiddle>
  );

  return (
    <Container>
      <RightCell>
        {time}
      </RightCell>
      <LeftCell isLastStop={isLastStop}>
        {renderCircle}
        {renderMiddleCircle}
        <Location isPassed={isLocationPassed}>{stop.location.crs}</Location>
        <TrainStateStyled>{trainState}</TrainStateStyled>
      </LeftCell>
    </Container>
  );
};

StopDetail.propTypes = {
  stop: PropTypes.object.isRequired,
  nextStop: PropTypes.object
};

export default StopDetail;
