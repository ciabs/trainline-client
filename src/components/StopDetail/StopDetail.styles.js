import styled from 'styled-components';

import colors from '../../styles/colors';
import config from '../../config';

export const Container = styled.div`
  display: flex;
  height: 50px;
`;

export const RightCell = styled.div`
  width: 60px;
`;

export const LeftCell = styled.div`
  border-left: ${props => props.isLastStop ? 'none' : `solid 1px ${colors.color3}`};
  padding-left: 20px;
  position: relative;
`;

export const Circle = styled.div`
  display: flex;
  position: absolute;
  border: solid 1px ${colors.color3};
  ${props => {
    switch (props.trainPosition) {
      case config.trainPosition.TRAIN:
        return `
        background-color: ${colors.color3};
        width: 20px;
        height: 20px;
        left: -11px;
        `;
      case config.trainPosition.EMPTY:
        return `
          background-color: white;
          width: 10px;
          height: 10px;
          left: -6px;
        `;
      case config.trainPosition.PASSED:
      default:
        return `
          background-color: black;
          width: 14px;
          height: 14px;
          left: -7px;
        `;
    }
  }}; 
  border-radius: 50%;
  align-items: center;
  justify-content: center;
`;

export const CircleMiddle = styled.div`
  display: flex;
  position: absolute;
  border: solid 1px ${colors.color3};
  ${props => {
    if (props.trainPosition === config.trainPosition.TRAIN) {
      return `
      background-color: ${colors.color3};
      width: 20px;
      height: 20px;
      left: -11px;
      top: 20px;
      `;
    }
  }}; 
  border-radius: 50%;
  align-items: center;
  justify-content: center;
`;

export const TrainStateStyled = styled.div`
  color: #999;  
  font-size: 14px;
`;

export const Location = styled.div`
  ${props => {
    if (props.isPassed) {
      return `
        color: rgb(128, 128, 128);
        font-weight: bold;
      `;
    }  
  }}
`;
