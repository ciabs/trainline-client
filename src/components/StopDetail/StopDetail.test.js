import React from 'react';
import ReactDom from 'react-dom';
import {shallow, configure} from 'enzyme';
import renderer from 'react-test-renderer';
import StopDetail from './StopDetail';
import Adapter from 'enzyme-adapter-react-16';
import moment from 'moment';

configure({ adapter: new Adapter() });

describe('StopDetail', () => {
  const props = {
    stop: {
      arrival: {
        scheduled: {
          scheduledTime: moment([2017,11,11,20,20,20])
        }
      },
      departure: {
        scheduled: {
          scheduledTime: moment([2017,11,11,20,20,20])
        }
      },
      location: {
        crs: 'London'
      }
    },
    nextStop: {}
  };

  it('should not crash when mounted', () => {
    shallow(
      <StopDetail
        {...props}
      />
    );
  });

  it('should not crash when mounted', () => {
    const div = document.createElement('div');
    ReactDom.render(
      <StopDetail
        {...props}
      />,
      div
    );
  });

  it('should have a valid snapshot', () => {
    const component = renderer.create(
      <StopDetail
        {...props}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
