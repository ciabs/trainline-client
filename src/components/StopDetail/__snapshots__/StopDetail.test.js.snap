// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopDetail should have a valid snapshot 1`] = `
<div
  className="sc-bdVaJa hUdZHX"
>
  <div
    className="sc-bwzfXH bsEiWU"
  >
    20:20
  </div>
  <div
    className="sc-htpNat bNmWsn"
  >
    <div
      className="sc-bxivhb dLUKcv"
    />
    <div
      className="sc-bZQynM boMXtV"
    >
      London
    </div>
    <div
      className="sc-EHOje hSaKJH"
    >
      On time
    </div>
  </div>
</div>
`;
